crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1631 @@
+            "parquet_source"
+        }
+
+        fn initialize(&mut self, _num_pipelines: usize) {
+            if self.verbose {
+                eprintln!("[ParquetSource]: {:?}", &self.config);
+            }
+
+            self.init_projected_arrow_fields();
+            self.physical_predicate = self.predicate.clone().map(phys_expr_to_io_expr);
+
+            let (row_group_rx, row_group_task_handle) = self.init_row_group_stream();
+
+            self.row_group_rx = Some(row_group_rx);
+            self.row_group_task_handle = Some(row_group_task_handle);
+        }
+
+        fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) {
+            debug_assert!(recv.is_empty());
+            debug_assert_eq!(send.len(), 1);
+
+            if self.is_finished {
+                send[0] = PortState::Done;
+                debug_assert!(self.row_group_rx.is_none());
+            } else if send[0] == PortState::Done {
+                {
+                    // Early shutdown - our port state was set to `Done` by the downstream nodes. This
+                    // also means we can discard any potential errors from the `shutdown()`. Note that
+                    // `shutdown()` internally unwraps any critical errors.
+                    let _ = polars_io::pl_async::get_runtime().block_on(self.shutdown());
+                };
+                self.is_finished = true;
+            } else {
+                debug_assert_eq!(send[0], PortState::Ready);
+            }
+        }
+
+        fn spawn<'env, 's>(
+            &'env mut self,
+            scope: &'s TaskScope<'s, 'env>,
+            recv: &mut [Option<RecvPort<'_>>],
+            send: &mut [Option<SendPort<'_>>],
+            _state: &'s ExecutionState,
+            join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+        ) {
+            debug_assert!(recv.is_empty());
+            debug_assert_eq!(send.len(), 1);
+
+            let mut morsel_tx = send[0].take().unwrap().serial();
+
+            self.row_group_data_task_init_permit.take();
+
+            join_handles.push(scope.spawn_task(TaskPriority::Low, async move {
+                let wait_group = WaitGroup::default();
+                let source_token = SourceToken::new();
+
+                let row_group_rx = self.row_group_rx.as_mut().unwrap();
+
+                loop {
+                    let Some(v) = row_group_rx.recv().await else {
+                        self.is_finished = true;
+                        // Shutdown to propagate any errors.
+                        self.shutdown().await?;
+                        break;
+                    };
+
+                    // Hold the permit until the downstream tasks are done with the morsel.
+                    let (handle, _row_group_decode_permit) = WithSemaphorePermit::split(v);
+                    let df = handle.await?;
+
+                    let seq = self.morsel_seq;
+                    self.morsel_seq = 1 + seq;
```

Comment:
Please store a `MorselSeq` and use `MorselSeq::successor`.

Reply:
added